opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp

```diff
@@ -613,8 +619,205 @@ class SimulatorFullyImplicitBlackoilEbos
+            }         
+            if( pu.phase_used[Gas]) {
+                //Gas phase rate per well
+                wr[2][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Gas]]);
+            }
+            //Unit conversion
+            if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_FIELD) { 
+                wr[0][w] = unit::convert::to(wr[0][w],unit::stb/unit::day);
+                wr[1][w] = unit::convert::to(wr[1][w],unit::stb/unit::day);
+                wr[2][w] = unit::convert::to(wr[2][w],1000*unit::cubic(unit::feet)/unit::day);
+                //BHP
+                wr[3][w] = unit::convert::to(xw.bhp()[w], unit::psia);
+                //THP
+                wr[4][w] = unit::convert::to(xw.thp()[w], unit::psia);
+            }
+            else if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_METRIC) { 
+                wr[0][w] = unit::convert::to(wr[0][w],unit::cubic(unit::meter)/unit::day);
+                wr[1][w] = unit::convert::to(wr[1][w],unit::cubic(unit::meter)/unit::day);
+                wr[2][w] = unit::convert::to(wr[2][w],unit::cubic(unit::meter)/unit::day);
+                //BHP
+                wr[3][w] = unit::convert::to(xw.bhp()[w], unit::barsa);
+                //THP
+                wr[4][w] = unit::convert::to(xw.thp()[w], unit::barsa);
+            }
+            else {
+                OPM_THROW(std::runtime_error, "Only 'FIELD' and 'METRIC' unit systems supported for well reporting");
+            }
+            
+            for (int n = 0; n < 3; n++){
+                wellRate += wr[n][w];
+            }
+            if (is_producer) {
+                if (pu.phase_used[Water] && pu.phase_used[Oil]) {
+                    if (wellRate > 0.00001){
+                        //Watercut
+                        wr[5][w] = wr[0][w]/wellRate;
+                    }
+                }
+                if (pu.phase_used[Oil] && pu.phase_used[Gas]) {
+                    //Gos-OilRatio (GOR)
+                    wr[6][w] = wr[2][w]/wr[1][w];
+                }
+                if (pu.phase_used[Water] && pu.phase_used[Gas]) {
+                    //Water-Gas Ratio (WGR)       
+                    wr[7][w] = wr[0][w]/wr[2][w];
+                }
+            }
+        }
+        return wr;
+    }
+   
+   //Compute and return produced and injected totals from wells per phase
+   std::vector<double> wellTotals(std::vector<std::vector<double>> wr,
+                                  const Wells* wells,
+                                  const Opm::PhaseUsage& pu)
+   {
+        int nw = wells->number_of_wells;
+        std::vector<double> wtotals (9,0.0);        
+        for (int w = 0; w < nw; ++w) { 
+           const bool is_producer = wells->type[w] == PRODUCER;
+           const bool is_injector = wells->type[w] == INJECTOR;
+           if (is_producer) {
+               //Field Oil produced
+               wtotals[0] += wr[1][w];
```

Comment:
I will extend this request to avoid hard-coded indices in general, using named constants instead, typically through an enum. Alternatively a struct with named fields (see for example the data::Well class) can be used instead of a flat array. (Sometimes you need a flat array for other reasons though.)